docs/notebooks/Writing_custom_interpreters_in_Jax.ipynb

```diff
@@ -0,0 +1,777 @@
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "foo\n",
+      "=====\n",
+      "invars: [a]\n",
+      "outvars: [b]\n",
+      "constvars: []\n",
+      "freevars: []\n",
+      "equation: [a, 1] add [b] {}\n",
+      "\n",
+      "jaxpr: { lambda  ;  ; a.\n",
+      "  let b = add a 1\n",
+      "  in [b] }\n",
+      "\n",
+      "\n",
+      "bar\n",
+      "=====\n",
+      "invars: [a, b, c]\n",
+      "outvars: [g, c]\n",
+      "constvars: [f]\n",
+      "freevars: []\n",
+      "equation: [a, c] dot_general [d] {'dimension_numbers': (((1,), (0,)), ((), ())), 'precision': None}\n",
+      "equation: [d, b] add [e] {}\n",
+      "equation: [e, f] add [g] {}\n",
+      "\n",
+      "jaxpr: { lambda f ;  ; a b c.\n",
+      "  let d = dot_general[ dimension_numbers=(((1,), (0,)), ((), ()))\n",
+      "                       precision=None ] a c\n",
+      "      e = add d b\n",
+      "      g = add e f\n",
+      "  in [g, c] }\n",
+      "\n"
+     ]
+    }
+   ],
+   "source": [
+    "def examine_jaxpr(jaxpr):\n",
+    "  print(\"invars:\", jaxpr.invars)\n",
+    "  print(\"outvars:\", jaxpr.outvars)\n",
+    "  print(\"constvars:\", jaxpr.constvars)\n",
+    "  print(\"freevars:\", jaxpr.freevars)\n",
+    "  for eqn in jaxpr.eqns:\n",
+    "    print(\"equation:\", eqn.invars, eqn.primitive, eqn.outvars, eqn.params)\n",
+    "  print()\n",
+    "  print(\"jaxpr:\", jaxpr)\n",
+    "\n",
+    "def foo(x):\n",
+    "  return x + 1\n",
+    "print(\"foo\")\n",
+    "print(\"=====\")\n",
+    "examine_jaxpr(jax.make_jaxpr(foo)(5))\n",
+    "\n",
+    "print()\n",
+    "\n",
+    "def bar(w, b, x):\n",
+    "  return np.dot(w, x) + b + np.ones(5), x\n",
+    "print(\"bar\")\n",
+    "print(\"=====\")\n",
+    "examine_jaxpr(jax.make_jaxpr(bar)(np.ones((5, 10)), np.ones(5), np.ones(10)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text",
+    "id": "k-HxK9iagnH6"
+   },
+   "source": [
+    "* `jaxpr.invars` - the `invars` of a Jaxpr are a list of the input variables to Jaxpr, analogous to arguments in Python functions\n",
+    "* `jaxpr.outvars` - the `outvars` of a Jaxpr are the variables that are returned by the Jaxpr. Every Jaxpr has multiple outputs.\n",
+    "* `jaxpr.constvars` - the `constvars` are a list of variables that are also inputs to the Jaxpr, but correspond to constants from the trace (we'll go over these in more detail later)\n",
+    "* `jaxpr.freevars` - these can arise when nesting `jit` and `pmap` transformations; we won't worry about them in this colab.\n",
+    "* `jaxpr.eqns` - a list of equations, which are essentially let-bindings. Each equation is list of input variables, a list of output variables, and a *primitive*, which is used to evaluate inputs to produce outputs. Each equation also has a set of `params`, a dictionary of keyword arguments to the primitive.\n",
```

Comment:
Last nit: a primitive's params aren't keyword arguments, since keyword arguments don't exist in jaxprs (that's a detail about how we embed primitive application in Python using `bind`). Parameters are essentially part of the primitive's name: you could imagine having one primitive `conv_padding=SAME` and another primitive `conv_padding=VALID`, but it's more convenient to think of those as part of the same `conv` family of primitives. That is, we index into parametric families of primitives by using `params` (together with a name for the family).

Those details aren't important to unpack, but we probably shouldn't say "keyword arguments" here.